src/network/network_crypto.cpp

```diff
@@ -0,0 +1,451 @@
+{
+	return std::make_unique<X25519EncryptionHandler>(this->derived_keys.ServerToClient(), this->nonce);
+}
+
+/**
+ * Read the key exchange data from a \c Packet that came from the client, and check whether the client passes the key
+ * exchange successfully.
+ * @param p The packet that has been received.
+ * @param derived_key_extra_payload The extra payload to pass to the key exchange.
+ * @return Whether the authentication was successful or not.
+ */
+NetworkAuthenticationServerHandler::ResponseResult X25519AuthenticationHandler::ReceiveResponse(Packet &p, std::string_view derived_key_extra_payload)
+{
+	if (p.RemainingBytesToTransfer() != X25519_KEY_SIZE + X25519_MAC_SIZE + X25519_KEY_EXCHANGE_MESSAGE_SIZE) {
+		Debug(net, 1, "[crypto] Received auth response of illegal size; authentication aborted.");
+		return NetworkAuthenticationServerHandler::NOT_AUTHENTICATED;
+	}
+
+	X25519KeyExchangeMessage message{};
+	X25519Mac mac;
+
+	p.Recv_bytes(this->peer_public_key);
+	p.Recv_bytes(mac);
+	p.Recv_bytes(message);
+
+	if (!this->derived_keys.Exchange(this->peer_public_key, X25519KeyExchangeSide::SERVER,
+			this->our_secret_key, this->our_public_key, derived_key_extra_payload)) {
+		Debug(net, 0, "[crypto] Peer sent an illegal public key; authentication aborted.");
+		return NetworkAuthenticationServerHandler::NOT_AUTHENTICATED;
+	}
+
+	if (crypto_aead_unlock(message.data(), mac.data(), this->derived_keys.ClientToServer().data(), nonce.data(),
+			this->peer_public_key.data(), this->peer_public_key.size(), message.data(), message.size()) != 0) {
+		/*
+		 * The ciphertext and the message authentication code do not match with the encryption key.
+		 * This is most likely an invalid password, or possibly a bug in the client.
+		 */
+		return NetworkAuthenticationServerHandler::NOT_AUTHENTICATED;
+	}
+
+	return NetworkAuthenticationServerHandler::AUTHENTICATED;
+}
+
+
+/* virtual */ NetworkAuthenticationClientHandler::RequestResult X25519PAKEClientHandler::ReceiveRequest(struct Packet &p)
+{
+	bool success = this->X25519AuthenticationHandler::ReceiveRequest(p);
+	if (!success) return NetworkAuthenticationClientHandler::INVALID;
+
+	this->handler->AskUserForPassword(this->handler);
+	return NetworkAuthenticationClientHandler::AWAIT_USER_INPUT;
+}
+
+/**
+ * Get the secret key from the given string. If that is not a valid secret key, reset it with a random one.
+ * Furthermore update the public key so it is always in sync with the private key.
+ * @param secret_key The secret key to read/validate/fix.
+ * @param public_key The public key to update.
+ * @return The valid secret key.
+ */
+/* static */ X25519SecretKey X25519AuthorizedKeyClientHandler::GetValidSecretKeyAndUpdatePublicKey(std::string &secret_key, std::string &public_key)
+{
+	X25519SecretKey key{};
+	if (!ConvertHexToBytes(secret_key, key)) {
+		Debug(net, 0, "[crypto] Found invalid secret key, creating a new random key");
```

Comment:
```suggestion
		if (!secret_key.empty()) {
			Debug(net, 0, "[crypto] Found invalid secret key, creating a new random key");
		}
```

At least, I assume that when the game is first started, the `secret_key` is still empty, and this will populate it for the first time. Which is a good thing. But that shouldn't show the message, I think :)